### ISSUE TITLE
inference with lora suddenly significantly slower

### ISSUE THREAD

**Issue by a-r-r-o-w** [huggingface/diffusers#12131]
inference with lora suddenly significantly slower than with diffusers==0.33.1 and peft==0.14.0
Used to be 3.9it/s now 2.9it/s

_Originally posted by @burgalon in https://github.com/huggingface/diffusers/pull/11999#issuecomment-3179413365_

**Comment by burgalon** [comment]
actually doesn't seem related to #11999 still trying to hunt the offending commit.
This happens when upgrading from 0.34.0 to dev

**Comment by burgalon** [comment]
@a-r-r-o-w seems like the offending PR is [#11180](https://github.com/huggingface/diffusers/pull/11180)
cbc8ced20f7a67254caac36af79e666ac2379833 is fast
0454fbb30bfbe21aa4ea29c827c396bac57dc518 is slow

**Comment by sayakpaul** [comment]
I wonder why it would make LoRA slower :/ FWIU, it doesn't have to do anything with PEFT or the LoRA related updates in `diffusers`? 

Also, if you could provide a short snippet on how you're benchmarking this, it would be quite helpful.

**Comment by burgalon** [comment]
sure
```
import torch
from diffusers import DiffusionPipeline

pipe = DiffusionPipeline.from_pretrained(
    '/black-forest-labs/FLUX.1-dev',
    torch_dtype=torch.bfloat16,
    local_files_only=True,
).to('cuda')

pipe.load_lora_weights('lora.safetensors', low_cpu_mem_usage=False)
pipe('test')
```

**Comment by sayakpaul** [comment]
Could we please keep this fully reproducible so that we can just essentially copy and paste the snippet for testing?

**Comment by burgalon** [comment]
does that work?
```
import torch
from diffusers import DiffusionPipeline
from huggingface_hub import hf_hub_download

pipe = DiffusionPipeline.from_pretrained(
    'black-forest-labs/FLUX.1-dev',
    torch_dtype=torch.bfloat16,
    local_files_only=True,
).to('cuda')
hf_hub_download(repo_id='burgalon/emma', filename='2910204.safetensors', local_dir='.')
pipe.load_lora_weights('./2910204.safetensors', low_cpu_mem_usage=False)
pipe('test')
# commit 0454fbb30bfbe21aa4ea29c827c396bac57dc518 SLOW 2.8it/s 9s
# commit cbc8ced20f7a67254caac36af79e666ac2379833 FAST 3.7it/s 7s
```

**Comment by burgalon** [comment]
@sayakpaul @a-r-r-o-w  i've narrowed down the issue to the changes in `src/diffusers/models/transformers/transformer_flux.py`
this revert resolves the issue https://github.com/burgalon/diffusers/commit/a401adb63a859f962215cdd3f2f2563f6b6a584d

**Comment by burgalon** [comment]
Commented https://github.com/huggingface/diffusers/pull/11180/files#r2273978819

Copilot :

This diff changes the way tensors are processed and returned in the forward pass of the transformer_flux model in the HuggingFace Diffusers library. The main performance issue comes from the repeated use of torch.cat and splitting tensors, which can introduce significant overhead, especially on large tensors or in high-throughput scenarios.

## Key Reasons for Slowness
1. Frequent Concatenation and Splitting
Before: hidden_states was used as a single tensor. Forward methods returned just hidden_states.
After:
At the start of each block, you concatenate encoder_hidden_states and hidden_states:
Python
text_seq_len = encoder_hidden_states.shape[1]
hidden_states = torch.cat([encoder_hidden_states, hidden_states], dim=1)
At the end of each block, you split them back:
Python
encoder_hidden_states, hidden_states = hidden_states[:, :text_seq_len], hidden_states[:, text_seq_len:]
Why it’s slow:
torch.cat allocates new memory and copies both tensors every time it’s called.
Slicing (hidden_states[:, :text_seq_len]) also allocates new memory for the split tensors.
If this is done in every block (as in your loop), the cost compounds quickly.
2. Increased Memory Footprint
Every concat/split operation increases peak memory usage, which can slow down training and inference (especially on GPU with limited bandwidth and memory).
3. Potential for Duplicate Computation
If encoder_hidden_states doesn’t change between blocks, you’re repeatedly concatenating the same tensor, which is wasteful.
Why Was This Change Made?
This pattern of concatenating then splitting often occurs in models that want cross-attention between different sources (e.g., text and image). It can make implementation simpler, but at the cost of speed and memory.

## How to Fix or Mitigate
Avoid repeated concat/split:
Concatenate only once per forward pass if possible, not inside every block.
Pass both tensors separately and only combine when strictly necessary for attention.
Profile with torch.autograd.profiler or nvprof:
Confirm that torch.cat and slicing are taking significant time.
Summary:
This diff causes slowness because every block in your transformer repeatedly concatenates and splits large tensors, leading to high memory usage and computational overhead. Optimize by minimizing these operations and only combining tensors when essential.

**Comment by a-r-r-o-w** [comment]
Thanks for investigating and finding the faulty commit @burgalon! When I tested the change during First Block Cache addition, the speed did not seem to be impacted. However, since you're noticing a difference, I will have to test again and we might have to revert the change. I suspect supporting first block cache easily with all models might require a significant rewrite to how it's implemented

**Comment by burgalon** [comment]
@a-r-r-o-w please note that it's not the FirstBlockCache but only this part
https://github.com/huggingface/diffusers/commit/37d88af22dc4779d5b92eb766bdaa2157f3dd7ae that's restoring performance

**Comment by a-r-r-o-w** [comment]
Yes, I understand. That change was made so that FBC becomes much simpler to implement. Without that change, FBC implementation becomes a very big mess of if-else statements

**Comment by burgalon** [comment]
correction - it actually seems like the problem is in `with self.transformer.cache_context("cond"):`
only reverting this https://github.com/burgalon/diffusers/commit/b07beebd42e8c45985bf550a01ec7c646ca6d8ef#diff-68d4ddab9d9846465f7d3b0a8583b256f7293df93e60a04eaeaced3b85de15aaL915-R946 resolves the slowness

**Comment by a-r-r-o-w** [comment]
@burgalon Just a small update.

I'm using the following code to benchmark:

```python
import time

import torch
from diffusers import FluxPipeline

pipe = FluxPipeline.from_pretrained("black-forest-labs/FLUX.1-dev", torch_dtype=torch.bfloat16)
pipe.to("cuda")
prompt = "A cat holding a sign that says hello world"
start_time = time.perf_counter()
latents = pipe(prompt, num_inference_steps=28, guidance_scale=4.0, generator=torch.Generator().manual_seed(42), output_type="latent").images
end_time = time.perf_counter()
elapsed_time = end_time - start_time
print(f"Elapsed time: {elapsed_time:.5f} seconds")
with torch.no_grad():
    latents = pipe._unpack_latents(latents, 1024, 1024, pipe.vae_scale_factor)
    latents = (latents / pipe.vae.config.scaling_factor) + pipe.vae.config.shift_factor
    image = pipe.vae.decode(latents, return_dict=False)[0]
image = pipe.image_processor.postprocess(image, output_type="pil")[0]
image.save("output.png")
```

With the cache context, I get:
```
Loading pipeline components...:   0%|                                       | 0/7 [00:00<?, ?it/s]You set `add_prefix_space`. The tokenizer needs to be converted from the slow tokenizers
Loading checkpoint shards: 100%|████████████████████████████████████| 2/2 [00:00<00:00, 40.49it/s]
Loading checkpoint shards: 100%|████████████████████████████████████| 3/3 [00:00<00:00, 18.14it/s]
Loading pipeline components...: 100%|███████████████████████████████| 7/7 [00:00<00:00,  7.50it/s]
100%|█████████████████████████████████████████████████████████████| 28/28 [00:15<00:00,  1.79it/s]
Elapsed time: 16.03116 seconds
```

Without the cache context changes, I get:
```
Loading pipeline components...:  14%|████▍                          | 1/7 [00:00<00:00,  7.03it/s]You set `add_prefix_space`. The tokenizer needs to be converted from the slow tokenizers
Loading checkpoint shards: 100%|████████████████████████████████████| 2/2 [00:00<00:00, 40.98it/s]
Loading checkpoint shards: 100%|████████████████████████████████████| 3/3 [00:00<00:00, 18.46it/s]
Loading pipeline components...: 100%|███████████████████████████████| 7/7 [00:00<00:00,  7.30it/s]
100%|█████████████████████████████████████████████████████████████| 28/28 [00:15<00:00,  1.80it/s]
Elapsed time: 15.98621 seconds
```

So, not really seeing a performance drop on that commit. My tests were on a single SXM4 A100 but I'll test on other devices as well.

**Comment by burgalon** [comment]
i was testing on h100. also the lora loading seems important to trigger the issue

**Comment by lauri9** [comment]
Hi there, I was profiling the Flux workload and noticed a slowdown similar as @burgalon . Please have a look at the screenshot of a zoomed in view of the profile trace attached below. The top part shows the CPU side with Python code, whereas the bottom part shows the GPU side with compute and communication kernels. The slowdown appears as a long blank space on the GPU side indicating idle time. This corresponds to exiting and entering the `cache_context` context manager on the CPU side.

<img width="2319" height="1044" alt="Image" src="https://github.com/user-attachments/assets/d8b149c6-efcb-476a-b969-4156e9353c15" />

I also tested with disabling the context manager by removing [this line](https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/flux/pipeline_flux.py#L943) after which the GPU kernels no longer have gaps between them.

It would seem that this regression is more likely to be visible on a fast GPU or a light workload, where the time spent in exiting/entering the cache context manager causes a break in work being sent to the GPU during which the GPU idles. Most of the time on the CPU side is spent checking torch version. The version checks were recently moved to cache in PR #12399 - after applying this PR the regression was fixed for me, and there are no more gaps on the GPU side. @burgalon could you try if this PR resolves the issue for you as well?

**Comment by cbensimon** [comment]
@lauri9 thanks a lot for confirming that versions caching mitigated the issue on your side! (cc @sayakpaul for visibility)

**Comment by sayakpaul** [comment]
Thanks to you for resolving that!

**Comment by yiyixuxu** [comment]
fixed in https://github.com/huggingface/diffusers/pull/12399
